File: ScheduleView/Database/WPF_VB/ScheduleViewRepository.vb
Public Class ScheduleViewRepository
	Private Shared m_context As ScheduleViewDBEntities
	Public Shared ReadOnly Property Context() As ScheduleViewDBEntities
		Get
			If m_context Is Nothing Then
				m_context = New ScheduleViewDBEntities()
			End If

			Return m_context
		End Get
	End Property

	Shared Sub New()
	End Sub

	Public Shared Function SaveData(action As Action) As Boolean
        Dim _IsSubmited = ScheduleViewRepository.Context.SaveChanges() >= 0
		If Not action Is Nothing AndAlso _IsSubmited Then
			action()
		End If
		Return _IsSubmited
	End Function

	Public Shared Function GetSqlAppointmentsByRange(start As DateTime, [end] As DateTime) As IQueryable(Of SqlAppointment)
		Dim ids = GetSqlAppointmentsIdsByRange(start, [end])

		Dim result = Context.SqlAppointments.Where(Function(a) ids.Contains(a.SqlAppointmentId)).ToList()

		' Load the recurrent appointments
		For Each item In Context.SqlAppointments.Where(Function(a) Not String.IsNullOrEmpty(a.RecurrencePattern))
			If RecurrenceHelper.IsOccurrenceInRange(item.RecurrencePattern, start, [end]) AndAlso Not result.Contains(item) Then
				result.Add(item)
			End If
		Next

		' Load the exceptions
		For Each item In Context.SqlAppointments.Where(Function(a) a.Start < [end] AndAlso a.SqlExceptionOccurrences.Count <> 0)
			If item.SqlExceptionOccurrences.Any(Function(e) e.SqlExceptionAppointment IsNot Nothing AndAlso e.SqlExceptionAppointment.Start >= start AndAlso e.SqlExceptionAppointment.[End] <= [end]) AndAlso Not result.Contains(item) Then
				result.Add(item)
			End If
		Next

		Return result.AsQueryable()
	End Function

	Private Shared Function GetSqlAppointmentsIdsByRange(start As DateTime, [end] As DateTime) As Integer()
		Dim result = Context.SqlAppointments.Where(Function(a) (a.Start >= start AndAlso a.[End] <= [end])).ToList()

		Return result.OfType(Of SqlAppointment)().[Select](Function(e) e.SqlAppointmentId).ToArray()
	End Function
End Class

File: ScheduleView/Database/WPF_CS/ScheduleViewRepository.cs
using System;
using System.Linq;
using ScheduleViewDB.Helpers;

namespace ScheduleViewDB
{
	public class ScheduleViewRepository
	{		
		private static ScheduleViewDBEntities context;
		public static ScheduleViewDBEntities Context
		{
			get
			{
				if (context == null)
				{
					context = new ScheduleViewDBEntities();
				}
				return context;
			}
		}

		public static bool SaveData(Action action)
		{
			var isSubmited = ScheduleViewRepository.Context.SaveChanges() >= 0;
			if (action != null && isSubmited)
			{
				action();
			}
			return isSubmited;
		}
		
		public static IQueryable<SqlAppointment> GetSqlAppointmentsByRange(DateTime start, DateTime end)
		{
			var ids = GetSqlAppointmentsIdsByRange(start, end);

			var result = context.SqlAppointments.Where(a => ids.Contains(a.SqlAppointmentId)).ToList<SqlAppointment>();

			// Load the recurrent appointments
			foreach (var item in context.SqlAppointments.Where(a => !string.IsNullOrEmpty(a.RecurrencePattern)))
			{
				if (RecurrenceHelper.IsOccurrenceInRange(item.RecurrencePattern, start, end) && !result.Contains(item))
				{
					result.Add(item);
				}
			}

			// Load the exceptions
			foreach (var item in context.SqlAppointments.Where(a => a.Start < end && a.SqlExceptionOccurrences.Count != 0))
			{
				if (item.SqlExceptionOccurrences.Any(e => e.SqlExceptionAppointment != null && e.SqlExceptionAppointment.Start >= start && e.SqlExceptionAppointment.End <= end) && !result.Contains(item))
				{
					result.Add(item);
				}
			}

			return result.AsQueryable<SqlAppointment>();
		}

		private static int[] GetSqlAppointmentsIdsByRange(DateTime start, DateTime end)
		{
            var result = context.SqlAppointments.Where(a => ((a.Start >= start && a.End <= end) || (a.Start <= start && a.End <= end) || (a.Start >= start && a.End >= end) || (a.Start <= start && a.End >= end))).ToList<SqlAppointment>();

			return result.OfType<SqlAppointment>().Select(e => e.SqlAppointmentId).ToArray();
		}
	}
}
